src/core/trulens/experimental/otel_tracing/core/instrument.py

```diff
@@ -0,0 +1,83 @@
+from functools import wraps
+import logging
+from typing import Any, Callable, Optional, Union
+
+from opentelemetry import trace
+from trulens.experimental.otel_tracing.core.init import TRULENS_SERVICE_NAME
+from trulens.experimental.otel_tracing.core.semantic import (
+    TRULENS_SELECTOR_NAME,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def instrument(
+    attributes: Optional[
+        Union[dict[str, Any], Callable[[Any, Any, Any], dict[str, Any]]]
+    ] = {},
+):
+    """
+    Decorator for marking functions to be instrumented in custom classes that are
+    wrapped by TruCustomApp, with OpenTelemetry tracing.
+    """
+
+    def _validate_selector_name(final_attributes: dict[str, Any]):
+        if TRULENS_SELECTOR_NAME in final_attributes:
+            selector_name = final_attributes[TRULENS_SELECTOR_NAME]
+            if not isinstance(selector_name, str):
+                raise ValueError(
+                    f"Selector name must be a string, not {type(selector_name)}"
+                )
+
+    def _validate_attributes(final_attributes: dict[str, Any]):
+        _validate_selector_name(final_attributes)
```

Comment:
also should validate that span types are a string too. Unless you plan on folding it into the TODO on L35.

Reply:
yeah I was thinking of folding it into the TODO - basically:
1. If the span type is provided, it must match one of the provided span types in the semcov package.
2. Otherwise it's unknown.